### PR TITLE
[dynamic shape] Add ObjectSpec container

### PR DESCRIPTION
**Pull request by fxdawnn**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #181961
* #181960
* __->__ #181841
* #181995

## Summary

Third in the dynamic-shape spec stack (after `IntSpec` and `TensorSpec`). Introduces `ObjectSpec` and extends `lookup_spec_from_dynamo_source` so the spec system can address tensors reached through Python attribute access (`obj.weight`, module params/buffers).

```python
torch.compile(
    fn,
    shapes_spec=ShapesSpec(
        params=ParamsSpec().arg(
            "model",
            ObjectSpec().field(
                "weight", TensorSpec([IntSpec.backed("h"), None])
            ),
        )
    ),
)
```

## What changed

- **`ObjectSpec`** — attribute-keyed container holding `{name: spec}` entries. Constructed fluently via `.field(name, spec)` or with a dict (`ObjectSpec({"weight": ...})`). Recursive: a value may be another container spec or a leaf (`IntSpec` / `TensorSpec` / `None`).
- **`lookup_spec_from_dynamo_source`** — extends the v0 `LocalSource` lookup to walk `Source` chains:
  - `LocalSource(name, is_input=True)` — the root of any walk
  - `AttrSource(base, member)` — descends into `ObjectSpec.field(member, ...)`
  - `NNModuleSource` — transparently unwrapped (it's a guard-semantics marker, not an access step)

  Other source kinds return `None` — extended in the DictSpec / ListSpec PRs that follow.

## Module-attribute access (`module.weight`)

`nn.Module` Parameter access goes through `NNModuleSource(AttrSource(...))` in dynamo, so the unwrap is what makes `ObjectSpec().field("weight", ...)` actually match the user's intent. Combined with the v0 spec-aware specialization fixes (`wrap_tensor` bypasses for spec'd tensors, `_automatic_dynamic` static-shape fast-path bypass), the spec drives the symbol creation for module attributes — including `nn.Parameter`.

## Test plan — `python test/dynamo/test_dynamic_spec.py`

- **`TestObjectSpec`** — empty / dict-form / fluent `.field` construction, `__setitem__`, `__contains__`, iter + items, repr (single / multiple / nested), recursive composition.
- **`TestObjectSpecLookup`** — `lookup_spec_from_dynamo_source` walks `AttrSource` chains against an `ObjectSpec`; nested-attr walk; missing-attr returns `None`; mismatched node kind returns `None`; bare `LocalSource` root preserves v0 behavior.
- **`TestObjectSpecCompile`** (`@skipIfTorchDynamo()`) — end-to-end: `shapes_spec=ShapesSpec(params=ParamsSpec().arg("obj", ObjectSpec().field("w", TensorSpec([IntSpec.backed("h"), None]))))`, compile `fn(obj)` with `obj.w` access, assert backed `SymInt` at the captured tensor placeholder's dim 0 and that varying dim-0 sizes don't recompile.

cc @ezyang @penguinwu @bobrenjc93 @aditvenk @laithsakka @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @azahed98

### CODE REVIEW

File: torch/_dynamo/dynamic_spec.py [hunk 1]
```diff
@@ -500,18 +504,196 @@ def __repr__(self) -> str:
...
+            return os
+        raise TypeError(
+            f"ObjectSpec.match cannot derive a spec for {type(obj).__name__}"
+        )
+
+    # No ``__eq__`` / ``__hash__``: same call as :class:`IntSpec` /
+    # :class:`TensorSpec`.
+
+
+# -- pytree registration -----------------------------------------------------
```
Comment (laithsakka) [inline torch/_dynamo/dynamic_spec.py]: why do u need this?!?

Comment (fxdawnn) [reply torch/_dynamo/dynamic_spec.py]: we don't. Removed and .match can be implemented some other ways.

File: test/dynamo/test_dynamic_spec.py [hunk 2]
```diff
@@ -954,5 +953,356 @@ def fn(p):
...
+        self.assertEqual(len(os), 0)
+
+    def test_dict_construction(self):
+        spec = IntSpec.backed("batch")
+        os = ObjectSpec({"x": spec, "n": IntSpec.static()})
+        self.assertEqual(len(os), 2)
+        self.assertIn("x", os)
+        self.assertIs(os["x"], spec)
+
+    def test_fluent_field(self):
```
Comment (laithsakka) [inline test/dynamo/test_dynamic_spec.py:970]: in the discussion today we decided to remove fluent API no need to add it for new ones.

File: torch/_dynamo/dynamic_spec.py [hunk 3]
```diff
@@ -500,18 +503,133 @@ def __repr__(self) -> str:
...
+
+    def items(self) -> Any:
+        return self._fields.items()
+
+    def __repr__(self) -> str:
+        entries = ", ".join(f".{name}: {spec!r}" for name, spec in self._fields.items())
+        return f"ObjectSpec({{{entries}}})"
+
+    @classmethod
+    def match(cls, obj: Any) -> Any:
```
Comment (laithsakka) [inline torch/_dynamo/dynamic_spec.py:598]: what is definition of match for you?
....
In my original proposal match takes an object and return bool
if the spec satisfy the object.

here you are using it to generate satic specs for all fields?

File: torch/_dynamo/dynamic_spec.py [hunk 4]
```diff
@@ -500,18 +503,133 @@ def __repr__(self) -> str:
...
         return None
-    # Only top-level function input args are supported for now.
-    #  Module attributes (self.x), globals, and values computed
-    #  during execution are not covered by shapes_spe yet.
-    if not isinstance(source, LocalSource) or not source.is_input:
+
+    # Walk source.base chain to its root, collecting path entries.
+    path: list[tuple[str, Any]] = []
+    cur = source
+    while not isinstance(cur, LocalSource):
```
Comment (laithsakka) [inline torch/_dynamo/dynamic_spec.py:528]: you did not add the test and fix for the parameter

**Review by laithsakka — CHANGES_REQUESTED** [review]

File: torch/_dynamo/dynamic_spec.py [hunk 5]
```diff
@@ -500,18 +503,133 @@ def __repr__(self) -> str:
...
+        ObjectSpec({"weight": TensorSpec([IntSpec.backed("h")])})
+
+    Recursion is allowed: a value may be another ``ObjectSpec`` or a
+    leaf (``IntSpec`` / ``TensorSpec``).
+    """
+
+    def __init__(self, fields: dict[str, Any] | None = None) -> None:
+        self._fields: dict[str, Any] = dict(fields or {})
+
+    def field(self, name: str, spec: Any) -> "ObjectSpec":
```
Comment (laithsakka) [inline torch/_dynamo/dynamic_spec.py:570]: lets remove fluent APIs from this new ones as i will remove them from all.

File: test/dynamo/test_dynamic_spec.py [hunk 6]
```diff
@@ -954,5 +953,356 @@ def fn(p):
...
+                super().__init__()
+                self.weight = torch.nn.Parameter(torch.randn(4, 3))
+
+        class Outer(torch.nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.inner = Inner()
+                self.bias = torch.nn.Parameter(torch.randn(4))
+
+        spec = ObjectSpec.match(Outer())
```
Comment (laithsakka) [inline test/dynamo/test_dynamic_spec.py:1296]: not what i thought match would do , do not see value for this function in its current form

**Review by laithsakka — CHANGES_REQUESTED** [review]

File: torch/_dynamo/dynamic_spec.py (hunk 3 above)
Comment (fxdawnn) [reply torch/_dynamo/dynamic_spec.py:598]: i see... i'll change it to match that requirements. Somehow i was thinking the ``.match`` would auto-parse the user input and return a specific (or even optimized) dynamic setup for users. (which might make their lives easier)

File: test/dynamo/test_dynamic_spec.py (hunk 6 above)
Comment (fxdawnn) [reply test/dynamo/test_dynamic_spec.py:1296]: ok...

**Comment by fxdawnn** [comment]
pivoted #182566
